src/transform/collect_free_vars.rs

```diff
@@ -0,0 +1,80 @@
+use std::collections::HashSet;
+
+use crate::destruct::{Destruct, Match};
+use crate::term::{SharedTerm, TermType};
+use crate::{
+    identifier::Ident,
+    term::{RichTerm, Term},
+};
+
+// `collect_free_vars` is transformation pass that caches a terms free variables.
+// This is used for optimizing memory consumption
+//
+// `free_vars` is the set of free variables in this term.
+//
+// `fields_free_vars` is a mapping between position of a field in a RecRecord`,
+//  and the free variables used in the body of the field.
+pub fn collect_free_vars(
+    rt: &mut RichTerm,
+    parent_rec_record: TermType,
+    free_vars: &mut HashSet<Ident>,
+    fields_free_vars: &mut Vec<HashSet<Ident>>,
+) {
+    if parent_rec_record == TermType::RecRecord {
+        fields_free_vars.push(free_vars.clone());
+    }
+
+    match SharedTerm::make_mut(&mut rt.term) {
+        Term::Var(id) => {
+            free_vars.insert(id.clone());
+        }
+        Term::Let(id, _, _, _) | Term::Fun(id, _) => {
```

Comment:
I thought it worked, because above the match on line 2, we store the set of free vars.
Since we store the set to the `RecRecord` before removing the binding.

```
    if parent_rec_record == TermType::RecRecord {
        fields_free_vars.push(free_vars.clone());
    }
```

Looking at it now I would expect it to fail with one more level of nesting, but it doesn't. 🤔
```
{
  x = 2,
  y = let z = 1 in let x = x + z in x,
}
```

I need to recheck my assumptions, and add tests.